project/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -62,47 +70,72 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-      }
-      Election electionData = new Election(name, date, candidateIds, candidateNames,
-                                           candidatePositions, candidatePartyAffiliation,
-                                           candidateIncumbency);
-      electionsData.add(electionData);
+      Entity election = electionQueryResult.asSingleEntity();
+      List<DirectoryCandidate> candidates =
+          extractCandidateInformation((List<String>) election.getProperty("candidateIds"),
+                                      (List<String>) election.getProperty("candidateIncumbency"));
+      List<Position> positions =
+          extractPositionInformation((List<String>) election.getProperty("candidatePositions"),
+                                     candidates);
+      electionsData.add(new Election(election.getKey().getName(),
+                                     (Date) election.getProperty("date"),
+                                     positions));
     }
     return electionsData;
   }
 
+  /**
+   * Queries the database for (brief version) candidate information based on ID and formats the
+   * data as {@code DirectoryCandidate} objects.
+   */
+  private List<DirectoryCandidate> extractCandidateInformation(List<String> candidateIds,
+      List<String> candidateIncumbency) {
+    List<DirectoryCandidate> candidates = new ArrayList<>(candidateIds.size());
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    for (int i = 0; i < candidateIds.size(); i++) {
+      String id = candidateIds.get(i);
+      String incumbency = candidateIncumbency.get(i);
+      Query candidateQuery = new Query("Candidate")
+          .setFilter(new FilterPredicate("__key__", FilterOperator.EQUAL,
+                      KeyFactory.createKey("Candidate", Long.parseLong(id))));
+      PreparedQuery candidateQueryResult = datastore.prepare(candidateQuery);
+      Entity candidate = candidateQueryResult.asSingleEntity();
+      candidates.add(new DirectoryCandidate(id, 
+                                            (String) candidate.getProperty("name"),
+                                            (String) candidate.getProperty("partyAffiliation"),
+                                            incumbency));
+    }
+    return candidates;
+  }
+
+  /**
+   * Formats position and its associated candidates' information as {@code Position}. Correlates
+   * one {@code Position} object with one or more {@code DierctoryCandidate}.
+   */
+  private List<Position> extractPositionInformation(List<String> candidatePositions,
+      List<DirectoryCandidate> candidates) {
+    Set<String> distinctPositions = new HashSet<>(candidatePositions);
+    List<Position> positions = new ArrayList<>(distinctPositions.size());
+    for (String positionName : distinctPositions) {
+      int startIndex = candidatePositions.indexOf(positionName);
+      int endIndex = candidatePositions.lastIndexOf(positionName);
```

Comment:
Are you using these anywhere?

Reply:
Ah thanks for the reminder! These should have been used to find the candidates corresponding to a particular position. I'll fix this